interface/default.py

```diff
@@ -0,0 +1,96 @@
+import dis
+import warnings
+
+from .compat import PY3
+from .formatting import bulleted_list
+from .functional import keysorted, sliding_window
+
+
+class default(object):
+    """Default implementation of a function in terms of interface methods.
+    """
+    def __init__(self, implementation):
+        self.implementation = implementation
+
+    def __repr__(self):
+        return "{}({})".format(type(self).__name__, self.implementation)
+
+
+class DefaultUsesNonInterfaceMembers(UserWarning):
+    pass
+
+
+if PY3:  # pragma: nocover-py2
+    _DEFAULT_USES_NON_INTERFACE_MEMBER_TEMPLATE = (
+        "Default implementation of {iface}.{method} uses non-interface attributes.\n\n"
+        "The following attributes may be accessed but are not part of "
+        "the interface:\n"
+        "{non_members}\n\n"
+        "Consider changing the implementation of {method} or making these attributes"
+        " part of {iface}."
+    )
+
+    def warn_if_defaults_use_non_interface_members(interface_name,
+                                                   defaults,
+                                                   members):
+        """Warn if an interface default uses non-interface members of self.
+        """
+        for method_name, attrs in non_member_attributes(defaults, members):
+            warnings.warn(_DEFAULT_USES_NON_INTERFACE_MEMBER_TEMPLATE.format(
```

Comment:
I wonder if we could have `default` take a list of names to ignore in this warning. One use case I imagine is a default implementation for `__init__` which will set some attribute, then another default implementation that uses that.

Reply:
I think if you're writing that pattern, then I think you just want an ABC, not an interface.  IMO it should always be safe to swap out the default implementation of a single method without having to know or care about the default implementation of any of the other methods.  If that's not the case, then I don't think the defaults are appropriate.

Reply:
Okay, that is reasonable.